### PR TITLE
Common exception changes

### PR DESCRIPTION
**Pull request by mattnayler**
### JIRA link (if applicable) ###

   [CPO-6](https://tools.hmcts.net/jira/browse/CPO-6) _"Build UpdatePaymentOrder endpoint"_
   [CPO-7](https://tools.hmcts.net/jira/browse/CPO-7) _"Build GetPaymentOrders endpoint"_

### Change description ###

Common exception changes extracted from CPO-6 (#48) and CPO-7 (#47)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
+    String CASE_ID_RG = "\\d{16}";
 
     default void buildErrors(final ConstraintValidatorContext context, final String message, List<String> errors) {
         final String errorsCommaSeparated = String.join(",", errors);
         context.buildConstraintViolationWithTemplate(message + errorsCommaSeparated + " are incorrect.")
             .addConstraintViolation();
+    }
+
+    static boolean isValidCaseId(String caseId) {
+
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java:21]: I would rather not make it static. I would avoid using statics.

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
+        return caseId != null
+            && caseId.matches(CASE_ID_RG)
+            && LUHN_CHECK_DIGIT.isValid(caseId);
+    }
 
+    static boolean isValidCpoId(String cpoId) {
+        try {
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java:28]: I would rather not make it static. I would avoid using statics.

File: src/test/java/uk/gov/hmcts/reform/BaseTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.time.LocalDateTime;
+import java.time.Month;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public interface BaseTest<T> {
+
```
Comment (Thor-tech-of-metal) [inline src/test/java/uk/gov/hmcts/reform/BaseTest.java:14]: The generic was removed  from the interface parameter as it only used for createInitialValuesList()..  createInitialValuesList() method signature can have the generic parameter.   I made this change in the my  PR.

File: src/main/java/uk/gov/hmcts/reform/cpo/security/AuthError.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.hmcts.reform.cpo.security;
+
+/**
+ * Authentication and Authorisation error messages.
+ **/
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/security/AuthError.java:5]: We can use lombok .  
I am not sure but lombok could gives us some facilities. for final classes with private constructors
@value(staticConstructor = "getInstance") for custom creation or others

**Review by Thor-tech-of-metal — COMMENTED** [review]
Matt please keep on eye my pr cos some changes has been done base on other reviews feedback.  I think that your pr should go first :) .  But keep updated with mine. 
We could do an end to end test too just in case .. we should not but it does not hurt :)

File: src/main/java/uk/gov/hmcts/reform/cpo/exception/ApiException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package uk.gov.hmcts.reform.cpo.exception;
+
+@SuppressWarnings({"PMD.MissingSerialVersionUID"})
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/cpo/exception/ApiException.java:3]: Minor: Can remove suppression as pmd has been removed (won't bother polluting the rest of the PR with the same comment, but same for elsewhere too!) 
```suggestion
```

File: src/main/java/uk/gov/hmcts/reform/cpo/service/impl/CasePaymentOrdersServiceImpl.java [hunk 6]
```diff
@@ -47,6 +50,12 @@ public CasePaymentOrdersServiceImpl(CasePaymentOrdersRepository casePaymentOrder
...
     }
 
+    @Transactional
+    @Override
+    public CasePaymentOrder updateCasePaymentOrder(UpdateCasePaymentOrderRequest request) {
+        throw new UnsupportedOperationException("Implement me: see CPO-6");
+    }
+
     private PageRequest getPageRequest(CasePaymentOrderQueryFilter casePaymentOrderQueryFilter) {
         final List<Sort.Order> orders = new ArrayList<>();
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/service/impl/CasePaymentOrdersServiceImpl.java:60]: getPageReques() has been moved to CasePaymentOrderQueryFilter, based on suggestions from reviewer.

**Review by danlysiak — APPROVED** [review]
Thanks!

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/IdValidator.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        context.buildConstraintViolationWithTemplate(String.join(", ", errors)).addConstraintViolation();
+        context.disableDefaultConstraintViolation();
+        return false;
+    }
+
+    @Override
+    public void validate(final String id, List<String> errors) {
+        if (!Validator.isValidCpoId(id)) {
+            errors.add(ValidationError.ID_INVALID);
+        }
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/validators/IdValidator.java:33]: ValidationError 👍  :)

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java [hunk 8]
```diff
@@ -1,16 +1,36 @@
...
         final String errorsCommaSeparated = String.join(",", errors);
         context.buildConstraintViolationWithTemplate(message + errorsCommaSeparated + " are incorrect.")
             .addConstraintViolation();
+    }
+
+    static boolean isValidCaseId(String caseId) {
+
+        return caseId != null
+            && caseId.matches(CASE_ID_RG)
+            && LUHN_CHECK_DIGIT.isValid(caseId);
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java:24]: LUHN_CHECK_DIGIT. 👍

**Review by sullivangareth — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java (hunk 1 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java]: Resolved in 94d7cc77493c1245220ac584184bb0099a9c3904

File: src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java (hunk 2 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/cpo/validators/Validator.java:27]: Resolved in 94d7cc77493c1245220ac584184bb0099a9c3904

File: src/test/java/uk/gov/hmcts/reform/BaseTest.java (hunk 3 above)
Comment (mattnayler) [reply src/test/java/uk/gov/hmcts/reform/BaseTest.java:14]: Resolved in 94d7cc77493c1245220ac584184bb0099a9c3904

File: src/main/java/uk/gov/hmcts/reform/cpo/security/AuthError.java (hunk 4 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/cpo/security/AuthError.java:5]: No adjustment made.  This class holds constatns only.  I reviewed switching to an enum but as this method of defining constants in classes has been used throughout CCD I have decided not to make the change.

File: src/main/java/uk/gov/hmcts/reform/cpo/service/impl/CasePaymentOrdersServiceImpl.java (hunk 6 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/cpo/service/impl/CasePaymentOrdersServiceImpl.java:60]: Thanks no action has been made in this PR.

File: src/main/java/uk/gov/hmcts/reform/cpo/exception/ApiException.java (hunk 5 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/cpo/exception/ApiException.java]: Resolved in 7546f595ec0be3b8185b0d483b205b48f1596a01

**Review by Thor-tech-of-metal — APPROVED** [review]

**Pull request merged**